### PR TITLE
feat(auth): JWT/OIDC authentication for MCP interactive endpoint (#1009)

### PR DESCRIPTION
**Pull request by jordigilh**
## Summary

- **JWT/OIDC authentication (Pattern B)**: Adds `JWTAuthenticator` with multi-issuer JWKS validation, `alg:none` rejection, JWKS pre-warm with 15s timeout, and graceful shutdown via `Close()`. `CompositeAuthenticator` routes JWT-first with K8s SA token fallback, fail-closed semantics.
- **main.go wiring + Helm**: Wires `CompositeAuthenticator` in kubernaut-agent startup. Fixes critical defer scope bug where `authCleanup` was called inside chi's route-setup closure (causing immediate JWKS context cancellation). Adds Helm chart `jwtProviders` config with schema validation.
- **Comprehensive test coverage**: 22 unit tests (JWT validation, composite routing, claims extraction, config validation), integration tests (full middleware pipeline), and 5 E2E tests with DEX OIDC provider deployed in-cluster (valid JWT, forged JWT rejection, Pattern A+B coexistence, tool invocation, token structure validation).
- **Architecture docs**: Updated DD-AUTH-MCP-001 v2.0 for dual-pattern auth. Added formal #1009 test plan with scenario IDs across all tiers.

Also includes prior commits: #1012 select_workflow internalization, #891 startup SAR self-check, and PR #1018 review feedback fixes.

## Business Requirements

- BR-INTERACTIVE-002: JWT/OIDC authentication for interactive MCP sessions
- BR-SECURITY-001: Token validation security (alg:none, JWKS poisoning)

## Test Plan

- [x] Unit tests: 22 scenarios pass (UT-KA-1009-001 through UT-KA-1009-008, composite, claims, config)
- [x] Integration tests: JWT middleware pipeline with RBAC
- [x] E2E tests: 5/5 pass with DEX OIDC in Kind cluster (E2E-KA-JWT-001 through 005)
- [x] Existing E2E suite: 81/81 pass (no regression)
- [x] Build: `go build ./...` clean
- [ ] Full CI pipeline validation

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by jordigilh — COMMENTED** [review]
## Multi-Persona Readiness Audit — PR #1022

**Audited by**: KA core team (automated multi-persona review)
**Scope**: Security, QE, Ops/Deployer, Product, Documentation

### Overall Verdict

| Persona | Verdict | Blockers |
|---------|---------|----------|
| **ProdSec** | ✅ PASS | 0 blocking (2 recommended hardenings — see inline) |
| **QE** | ⚠️ CONDITIONAL | TEST_PLAN drift: missing ITs, path mismatches (see inline + general findings below) |
| **Product** | ✅ PASS | BR-INTERACTIVE-002 + BR-SECURITY-001 addressed |
| **Operator/Deployer** | ✅ PASS | 0 (2 caveats for JWT-enabled deployments — see inline) |
| **Release Engineering** | ✅ PASS | Clean build, 81/81 existing E2E + 5 new JWT E2E |
| **Documentation** | ⚠️ CONDITIONAL | DD timeout/failure mode drift vs code (see inline) |
| **SRE** | ✅ PASS | ProviderType for observability, /readyz coverage |

---

### QE-3 [CONDITIONAL]: TEST_PLAN drift — missing integration tests

`docs/tests/1009/TEST_PLAN.md` lists the following integration test IDs as **Pending**, but they have **no matching test code** in the repo:

- **IT-KA-1009-003**: JWT user passes SAR check → not implemented
- **IT-KA-1009-006**: ProviderType in audit logs → not implemented
- **IT-KA-1009-008**: JWKS pre-warm failure → soft-disable → not implemented (and conflicts with actual code behavior — see DOC-3 below)
- **IT-KA-1009-009**: Helm ClusterRoleBinding group → not implemented

Additionally, the TEST_PLAN references file paths like `test/integration/kubernautagent/auth/mixed_traffic_test.go` and `test/integration/kubernautagent/auth/`, but the actual integration tests live in `test/integration/kubernautagent/jwt_middleware_test.go`.

**Action**: Either implement the missing ITs or update the TEST_PLAN to remove/defer them. Update file paths to match reality.

---

### QE-4: E2E-KA-JWT-005 does not validate expiry rejection

The test description says "token structure validation" but doesn't actually test expired-token rejection from a real IdP (acknowledged in-file comment). Expiry is covered by UT + IT tiers, but the E2E name may mislead QE. Consider renaming for clarity.

---

### SEC-7 [Low]: JWT errors → 500 vs 401 semantics

In `pkg/shared/auth/middleware.go`, only `errors.Is(err, ErrTokenInvalid)` maps to **401**. Other JWT failures (JWKS fetch errors, claim extraction failures) return **500** ("Authentication service unavailable"). This is fail-closed for access (good), but blurs client-visible semantics. Acceptable for v1.5 — consider refining error mapping for v1.6.

File: pkg/shared/auth/jwt_auth.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	keyset, err := provider.jwksCache.Lookup(ctx, provider.entry.JWKSURL)
+	if err != nil {
+		return UserInfo{}, fmt.Errorf("fetching JWKS for issuer %q: %w", issuer, err)
+	}
+
+	verifiedToken, err := jwt.Parse(
+		[]byte(rawToken),
+		jwt.WithKeySet(keyset, jws.WithInferAlgorithmFromKey(true)),
+		jwt.WithAudience(provider.entry.Audience),
+		jwt.WithValidate(true),
```
Comment (jordigilh) [inline pkg/shared/auth/jwt_auth.go:156]: **SEC-1 [Recommended hardening]**: Defense-in-depth — consider adding explicit `jwt.WithIssuer(provider.entry.Issuer)` here.

Currently, `iss` is used unverified for routing (`extractUnverifiedIssuer`), then signature verification happens against that provider's JWKS. This is safe because changing `iss` breaks the signature. However, explicitly re-validating `iss` against the configured provider would add an extra layer and align with OIDC best practices.

Not blocking — signature verification already prevents cross-issuer attacks, and unit test `UT-KA-1009-013` validates this.

File: pkg/shared/auth/jwt_auth.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	username, err := ExtractStringClaim(claims, provider.entry.UsernameClaim)
+	if err != nil {
+		return UserInfo{}, fmt.Errorf("extracting username from claim %q: %w", provider.entry.UsernameClaim, err)
+	}
+
+	groups, err := ExtractGroupsClaim(claims, provider.entry.GroupsClaim)
+	if err != nil {
+		groups = []string{}
```
Comment (jordigilh) [inline pkg/shared/auth/jwt_auth.go:176]: **SEC-5 [Low]**: Silent fallback to empty groups on extraction failure.

If `ExtractGroupsClaim` fails (wrong type, missing claim), the user gets `[]string{}`. This means authorization that depends **only** on group membership must compensate — if no groups are required for the SAR check, this is fine. But if a downstream RBAC gate expects specific groups, a user with a malformed groups claim could silently lose group-based privileges without an error surfacing.

Consider logging at `V(1)` when groups extraction fails, so operators can diagnose identity issues. Same pattern as the enrichment skip log from PR #1018.

File: cmd/kubernautagent/main.go [hunk 3]
```diff
@@ -1097,23 +1111,56 @@ func (f *dsCatalogFetcher) FetchValidator(ctx context.Context) (*parser.Validato
...
+				JWKSURL:       p.JWKSURL,
+				Audience:      p.Audience,
+				UsernameClaim: p.ClaimMappings.Username,
+				GroupsClaim:   p.ClaimMappings.Groups,
+			}
+		}
+
+		jwtAuth, err := auth.NewJWTAuthenticator(entries)
+		if err != nil {
+			logger.Error(err, "failed to create JWTAuthenticator; Pattern B disabled, Pattern A active")
```
Comment (jordigilh) [inline cmd/kubernautagent/main.go:1151]: **SEC-6 / DOC-3 [Medium]**: Global Pattern B disable vs DD-AUTH-MCP-001 "per-provider" claim.

`NewJWTAuthenticator` takes **all** provider entries and returns a single error if **any** provider's JWKS pre-warm fails. This means one unreachable JWKS endpoint disables JWT auth **globally** — all providers, not just the failing one.

DD-AUTH-MCP-001 v2.0 (line 243) states: _"JWKS pre-warm failed; Pattern B disabled **for this provider**"_ and _"JWKS cache auto-refreshes on subsequent requests"_ — implying per-provider degradation with runtime recovery. The actual code behavior is:

1. **Global** disable (all providers) on any single pre-warm failure
2. **No runtime recovery** — construction fails, `CompositeAuthenticator` is never created
3. Pattern A continues to work (good)

Options:
- **A)** Update DD-AUTH-MCP-001 to match code (global disable on any pre-warm failure)
- **B)** Implement per-provider skip in `NewJWTAuthenticator` (skip failing provider, register others)

For v1.5 with a single provider (Keycloak), the behavior is equivalent. This becomes material in v1.6 with SPIRE as a second provider.

File: internal/kubernautagent/config/config.go [hunk 4]
```diff
@@ -208,12 +209,102 @@ type MCPConfig struct {
...
-	RateLimitPerUser      int           `yaml:"rateLimitPerUser"`
-	MaxAnalyzingTimeout   time.Duration `yaml:"maxAnalyzingTimeout"`
+	Enabled               bool                `yaml:"enabled"`
+	SessionTTL            time.Duration       `yaml:"sessionTTL"`
+	InactivityTimeout     time.Duration       `yaml:"inactivityTimeout"`
+	MaxConcurrentSessions int                 `yaml:"maxConcurrentSessions"`
+	RateLimitPerUser      int                 `yaml:"rateLimitPerUser"`
+	MaxAnalyzingTimeout   time.Duration       `yaml:"maxAnalyzingTimeout"`
+	JWTProviders          []JWTProviderConfig `yaml:"jwtProviders,omitempty"`
+	JWTInteractiveGroup   string              `yaml:"jwtInteractiveGroup,omitempty"`
```
Comment (jordigilh) [inline internal/kubernautagent/config/config.go]: **OPS-4 [Low]**: `JWTInteractiveGroup` is defined in config, Helm values, and schema, but **not referenced anywhere in the agent binary**.

`grep -r "JWTInteractiveGroup\|jwtInteractiveGroup" cmd/ internal/ pkg/` shows zero references outside config struct definition and test files. Operators may configure this field expecting it to enforce group-based access to interactive mode, but it currently has no effect.

Either:
- Wire it into the MCP auth middleware (if the intent is to gate interactive access by group)
- Remove it from config/Helm/schema to avoid false confidence
- Document it explicitly as "reserved for v1.6" if it's a placeholder

File: docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md [hunk 5]
```diff
@@ -177,11 +230,28 @@ for key := range r.Header {
...
 ```
 
-MCP auth handler reads from a **preserved copy** (captured before stripping) only after SAR verification for Pattern B.
+Pattern B extracts identity from **verified JWT claims**, not from HTTP headers. The header stripping defense layer remains active for both patterns.
+
+### JWKS Pre-warm at Startup
+
+```go
+for _, provider := range cfg.Interactive.JWTProviders {
+    ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (jordigilh) [inline docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md]: **DOC-2 [Medium]**: Timeout mismatch — DD says 10s, code says 15s.

The actual `jwksPreWarmTimeout` constant in `pkg/shared/auth/jwt_auth.go:74` is `15 * time.Second`, not `10*time.Second` as shown in this pseudocode block.

Additionally, the DD code example uses a `PreWarm` method that doesn't exist in the implementation — the actual pre-warm happens inside `NewJWTAuthenticator` via `c.Lookup(warmCtx, e.JWKSURL)`.

This section, the Risks table (line 332: "PreWarm with 10s timeout"), and the Consequences section (line 322: "Pre-warm with 10s timeout") should all be updated to reflect `15s` and the actual construction-time behavior.

File: docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md [hunk 6]
```diff
@@ -177,11 +230,28 @@ for key := range r.Header {
...
+    err := jwtAuth.PreWarm(ctx, provider.Issuer)
+    cancel()
+    if err != nil {
+        logger.Error(err, "JWKS pre-warm failed; Pattern B disabled for this provider",
```
Comment (jordigilh) [inline docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md]: **DOC-3 [Medium]**: Per-provider disable claim contradicts actual behavior.

This pseudocode says _"Pattern B disabled **for this provider**"_ with a comment _"Pattern B returns 401 for this issuer"_, implying per-provider degradation. The actual code in `cmd/kubernautagent/main.go:1142-1144` takes a different path:

```go
jwtAuth, err := auth.NewJWTAuthenticator(entries)
if err != nil {
    logger.Error(err, "failed to create JWTAuthenticator; Pattern B disabled, Pattern A active")
}
```

If **any** provider's pre-warm fails, `NewJWTAuthenticator` returns an error and the **entire** `CompositeAuthenticator` is not created — Pattern B is disabled **globally**, not per-provider.

For v1.5 (single provider) this is academic. For v1.6 (Keycloak + SPIRE), this section should be updated to match actual semantics or the code should be changed to support per-provider degradation.

File: docs/tests/1009/TEST_PLAN.md [hunk 7]
```diff
@@ -0,0 +1,784 @@
...
+- **Behavior**: Fail-closed prevents cross-path security leak
+- **Correctness**: Known issuer errors are terminal, no fallback
+- **Security**: Attacker cannot forge a JWT with known issuer to bypass into K8s TokenReview path
+
+### IT-KA-1009-005: Mixed Pattern A + Pattern B traffic
+
+**BR**: BR-INTERACTIVE-001, BR-INTERACTIVE-002
+**Priority**: P0
+**Type**: Integration
+**File**: `test/integration/kubernautagent/auth/mixed_traffic_test.go`
```
Comment (jordigilh) [inline docs/tests/1009/TEST_PLAN.md]: **QE-3 [Conditional]**: This file path does not exist.

The TEST_PLAN references `test/integration/kubernautagent/auth/mixed_traffic_test.go` and `test/integration/kubernautagent/auth/` (also at lines 453, 678, 694, 702), but the actual integration tests live at:

```
test/integration/kubernautagent/jwt_middleware_test.go
```

Additionally, several integration test IDs listed in the plan have no matching implementation:
- `IT-KA-1009-003` (JWT user passes SAR) — not implemented
- `IT-KA-1009-006` (ProviderType in audit logs) — not implemented
- `IT-KA-1009-008` (JWKS pre-warm failure) — not implemented (and conflicts with actual code behavior — see DOC-3 comment on DD)
- `IT-KA-1009-009` (Helm ClusterRoleBinding) — not implemented

Please either implement the missing tests or update the plan to defer/remove them and correct the file paths.

File: internal/kubernautagent/config/config.go [hunk 8]
```diff
@@ -208,12 +209,102 @@ type MCPConfig struct {
...
+
+// validateJWKSURL checks that the JWKS URL is syntactically valid and uses
+// an HTTP or HTTPS scheme. HTTPS enforcement in production is delegated to
+// the kubernaut-operator admission webhook (kubernaut-operator#46).
+func validateJWKSURL(rawURL string) error {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return fmt.Errorf("invalid jwksURL %q: %w", rawURL, err)
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
```
Comment (jordigilh) [inline internal/kubernautagent/config/config.go:303]: **SEC-4 [Medium — accepted risk]**: HTTPS-only JWKS not enforced in KA.

`http://` JWKS URLs are accepted here. In a production environment, a plain HTTP JWKS endpoint is vulnerable to MITM attacks — an attacker on the network path could serve forged JWKS keys and sign arbitrary JWTs.

The comment at line 299 documents this is deferred to `kubernaut-operator#46` (admission webhook). This is an acceptable risk posture for v1.5 if the operator enforces HTTPS, but:

1. Ensure `kubernaut-operator#46` is tracked and prioritized before any non-dev deployment
2. Consider adding a startup **warning log** when `http://` JWKS URLs are configured (similar to how K8s API server warns about insecure backends)

File: test/unit/shared/auth/jwt_auth_test.go [hunk 9]
```diff
@@ -0,0 +1,335 @@
+/*
```
Comment (jordigilh) [inline test/unit/shared/auth/jwt_auth_test.go:1]: **QE-5/6/7/8 [Low — recommended additions]**: Missing edge-case test coverage.

The core JWT validation paths are well covered (signature, expiry, audience, alg:none, multi-issuer). However, the following edge cases are not exercised:

1. **QE-5: JWKS runtime failure** — JWKS server goes down _after_ successful pre-warm (mid-operation). Currently only pre-warm failure at construction time is tested. Consider a test where MockJWKSServer is shut down between construction and token validation.

2. **QE-6: Concurrency / race condition** — No parallel `ValidateTokenFull` stress test. Consider adding a test with `N` goroutines validating tokens concurrently against one `JWTAuthenticator` (run with `-race`).

3. **QE-7: `nbf` (not-before) / clock skew** — Expiry is tested, but `nbf` claim enforcement is not exercised. If `jwt.WithValidate(true)` checks `nbf`, it should be tested.

4. **QE-8: `aud` as array** — Keycloak can send `aud` as `["aud1", "aud2"]`. MockJWKSServer issues tokens with a single-string audience. Add a test with array `aud` to confirm the `jwx` library handles it correctly with `jwt.WithAudience()`.

None of these are blocking, but they strengthen the security test posture for v1.6 multi-provider rollout.

File: test/unit/shared/auth/claims_extraction_poc_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+	Describe("PoC-CLAIMS-003: Nested claim — realm_access.roles", func() {
+		It("should extract roles from nested realm_access", func() {
+			val, err := auth.ExtractGroupsClaim(keycloakStandard, "realm_access.roles")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(val).To(ConsistOf("offline_access", "uma_authorization", "kubernaut-user"))
+		})
+	})
+
+	Describe("PoC-CLAIMS-004: Double-nested claim — resource_access.kubernaut-agent.roles", func() {
```
Comment (jordigilh) [inline test/unit/shared/auth/claims_extraction_poc_test.go:57]: **QE-9 [Low]**: PoC test ID vs TEST_PLAN ID mismatch.

This test (`PoC-CLAIMS-004`) covers the double-nested Keycloak claim extraction (`resource_access.<client>.roles`), which is exactly what TEST_PLAN lists as `UT-KA-1009-004`. However, the BR matrix in `docs/tests/1009/TEST_PLAN.md` expects `UT-KA-1009-004` to be a full `JWTAuthenticator.ValidateTokenFull` end-to-end test with double-nested groups — not just the extraction layer.

Options:
- **A)** Promote this PoC to a formal UT by adding a `JWTAuthenticator` test that uses double-nested Keycloak claims end-to-end (through `ValidateTokenFull`, not just `ExtractGroupsClaim`)
- **B)** Update TEST_PLAN to accept `PoC-CLAIMS-004` as coverage for `UT-KA-1009-004`

**Comment by jordigilh** [comment]
## Review Response — Batch 1 (7 items addressed)

Pushed `5c4c8b90b` addressing the following findings:

### Code fixes
| ID | Finding | Resolution |
|----|---------|------------|
| **SEC-1** | Add `jwt.WithIssuer()` defense-in-depth | Added — issuer now re-validated during `jwt.Parse` |
| **SEC-5** | Silent groups fallback | Added `V(1)` log with username, claim path, and error. `NewJWTAuthenticator` now accepts `logr.Logger` |
| **OPS-4** | Dead `JWTInteractiveGroup` field | Removed from config, Helm template, values, schema, and all test files |

### Documentation fixes
| ID | Finding | Resolution |
|----|---------|------------|
| **DOC-2** | Timeout 10s → 15s, non-existent `PreWarm()` | Updated pseudocode to match actual `NewJWTAuthenticator` construction + 15s constant |
| **DOC-3 / SEC-6** | Per-provider vs global disable | Updated DD to document global disable semantics. Noted per-provider degradation for v1.6 |
| **QE-3** | TEST_PLAN paths + missing ITs | Fixed file paths to `jwt_middleware_test.go`. Deferred IT-003/006/008/009 to v1.6 |

### Deferred to v1.6
| ID | Finding | Rationale |
|----|---------|-----------|
| **SEC-7** | JWT errors → 500 vs 401 refinement | Fail-closed is correct; semantics refinement for v1.6 |
| **QE-4** | E2E-KA-JWT-005 naming | Cosmetic; expiry covered by UT + IT tiers |

### Verification
- `go build ./...` — clean
- Unit tests: 22 JWT + 109 config — all pass
- Integration tests: 5 JWT middleware — all pass

**Comment by jordigilh** [comment]
## Batch 2: Comments 8-10 resolved (af2560ece)

### SEC-4 — HTTP JWKS warning log
Added startup warning when JWT providers use plain `http://` JWKS URLs:
```
WARNING: JWKS URL uses plain HTTP — vulnerable to MITM in production;
enforce HTTPS via kubernaut-operator admission webhook (kubernaut-operator#46)
```
Logged with `provider` and `jwksURL` structured fields. HTTPS enforcement deferred to operator admission webhook as documented.

### QE-5/6/7/8 — 4 edge-case tests added
All run with `-race` enabled:
| ID | Test | Result |
|----|------|--------|
| QE-5 | JWKS cache resilience — cached keys survive server shutdown | ✅ |
| QE-6 | Concurrent `ValidateTokenFull` (20 goroutines, `-race`) | ✅ |
| QE-7 | `nbf` (not-before) enforcement — future `nbf` rejected | ✅ |
| QE-8 | `aud` as array — Keycloak multi-audience tokens accepted | ✅ |

### QE-9 / UT-KA-1009-004 — Full ValidateTokenFull with double-nested claims
Added a full end-to-end `ValidateTokenFull` test using `resource_access.kubernaut-agent.roles` (Keycloak double-nested claim path). Validates username, groups extraction, and `ProviderType` through the complete authenticator pipeline. This satisfies `UT-KA-1009-004` from the TEST_PLAN.

**All 27 JWT tests pass. Build clean. All 10 review comments now addressed.**

File: pkg/shared/auth/jwt_auth.go (hunk 1 above)
Comment (jordigilh) [reply pkg/shared/auth/jwt_auth.go:156]: Done in 5c4c8b9. Added `jwt.WithIssuer(provider.entry.Issuer)` to `jwt.Parse` options in `ValidateTokenFull` — defense-in-depth issuer re-validation now happens alongside signature verification.

File: pkg/shared/auth/jwt_auth.go (hunk 2 above)
Comment (jordigilh) [reply pkg/shared/auth/jwt_auth.go:176]: Done in 5c4c8b9. Added `a.logger.V(1).Info` when `ExtractGroupsClaim` fails, logging `username`, `groupsClaim`, and `error` — provides operational observability without changing the fallback behavior (empty groups still returned).

File: cmd/kubernautagent/main.go (hunk 3 above)
Comment (jordigilh) [reply cmd/kubernautagent/main.go:1151]: Done in 5c4c8b9. Updated DD-AUTH-MCP-001 to accurately reflect actual behavior: `NewJWTAuthenticator` performs a synchronous 15s-timeout pre-warm, and failure disables Pattern B **globally** (not per-provider). Removed stale pseudocode and linked to actual code. Updated Consequences and Risks sections to match.

File: internal/kubernautagent/config/config.go (hunk 4 above)
Comment (jordigilh) [reply internal/kubernautagent/config/config.go]: Done in 5c4c8b9. Removed `JWTInteractiveGroup` as dead code from all locations:
- `internal/kubernautagent/config/config.go` — field removed from `InteractiveConfig`
- `charts/kubernaut/templates/kubernaut-agent/kubernaut-agent.yaml` — template block removed
- `charts/kubernaut/values.schema.json` — property removed
- `charts/kubernaut/values.yaml` — commented definition removed
- `test/unit/kubernautagent/config/jwt_config_test.go` — assertion removed
- `test/infrastructure/kubernautagent.go` — ConfigMap reference removed

File: docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md (hunk 5 above)
Comment (jordigilh) [reply docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md]: Done in 5c4c8b9. Fixed the timeout mismatch: updated DD to reflect the actual `jwksPreWarmTimeout = 15 * time.Second` constant. Removed the stale `PreWarm` pseudocode and replaced with a description of the actual `c.Lookup(warmCtx, e.JWKSURL)` pre-warm that happens inside `NewJWTAuthenticator`. Also updated the Risks table reference from "10s" to "15s".

File: docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md (hunk 6 above)
Comment (jordigilh) [reply docs/architecture/decisions/DD-AUTH-MCP-001-mcp-endpoint-security.md]: Done in 5c4c8b9. Updated DD-AUTH-MCP-001 pseudocode and prose to state Pattern B is disabled **globally** (not per-provider) when any JWKS pre-warm fails. This accurately reflects the actual code path where `NewJWTAuthenticator` returns a single error and the caller falls back to Pattern A only.

File: docs/tests/1009/TEST_PLAN.md (hunk 7 above)
Comment (jordigilh) [reply docs/tests/1009/TEST_PLAN.md]: Done in 5c4c8b9. Fixed all file path references from `test/integration/kubernautagent/auth/mixed_traffic_test.go` to the actual location `test/integration/kubernautagent/jwt_middleware_test.go`. Deferred 4 unimplemented integration tests (IT-KA-1009-003, 006, 008, 009) to v1.6 with rationale documented in the plan. Updated BR Coverage Matrix and Test Scenarios sections accordingly.

File: internal/kubernautagent/config/config.go (hunk 8 above)
Comment (jordigilh) [reply internal/kubernautagent/config/config.go:303]: Done in af2560e. Added a startup warning log when `http://` JWKS URLs are configured:

```
WARNING: JWKS URL uses plain HTTP — vulnerable to MITM in production;
enforce HTTPS via kubernaut-operator admission webhook (kubernaut-operator#46)
```

Logged with `provider` (issuer) and `jwksURL` structured fields. HTTPS enforcement remains deferred to `kubernaut-operator#46` (admission webhook) as accepted risk for v1.5.

File: test/unit/shared/auth/jwt_auth_test.go (hunk 9 above)
Comment (jordigilh) [reply test/unit/shared/auth/jwt_auth_test.go:1]: Done in af2560e. All 4 edge-case tests added and passing with `-race`:

| ID | Test | Notes |
|----|------|-------|
| QE-5 | JWKS cache resilience | Validates cached keys survive server shutdown (graceful degradation) |
| QE-6 | Concurrent validation | 20 goroutines × `ValidateTokenFull`, run with `-race` detector |
| QE-7 | `nbf` enforcement | Future `nbf` rejected as `ErrTokenInvalid` via `jwt.WithValidate(true)` |
| QE-8 | `aud` as array | Multi-audience `["kubernaut-agent", "another-service"]` accepted correctly |

File: test/unit/shared/auth/claims_extraction_poc_test.go (hunk 10 above)
Comment (jordigilh) [reply test/unit/shared/auth/claims_extraction_poc_test.go:57]: Done in af2560e — went with Option A. Added a formal `ValidateTokenFull` test (`UT-KA-1009-004`) that exercises the full authenticator pipeline with double-nested Keycloak claims (`resource_access.kubernaut-agent.roles`). The test validates username, groups extraction, and `ProviderType` end-to-end through `ValidateTokenFull`, not just the extraction layer. The existing `PoC-CLAIMS-004` remains as extraction-layer coverage.

**Comment by jordigilh** [comment]
## Response to Multi-Persona Readiness Audit (Review)

All 10 inline findings have been resolved in commits `5c4c8b9` (batch 1) and `af2560e` (batch 2). Inline replies posted on each comment thread.

Addressing the top-level findings:

### QE-3 — TEST_PLAN drift
**Resolved** in `5c4c8b9`. File paths corrected to `test/integration/kubernautagent/jwt_middleware_test.go`. Four unimplemented ITs (IT-KA-1009-003, 006, 008, 009) deferred to v1.6 with rationale documented in the plan.

### QE-4 — E2E-KA-JWT-005 naming
**Acknowledged**. The test validates token structure (format, claims presence) rather than expiry rejection — expiry is covered at UT/IT tiers. Agree the name could be clearer; will consider renaming in a follow-up if it causes confusion.

### SEC-7 — JWT errors 500 vs 401 semantics
**Acknowledged as accepted for v1.5**. The fail-closed behavior (non-`ErrTokenInvalid` → 500) is intentional to avoid leaking internal failure details. Agree that refining error mapping (e.g., distinguishing JWKS-fetch transient errors from auth failures) is a good v1.6 improvement.

---

**Summary**: All blocking and conditional items resolved. QE-4 and SEC-7 tracked for v1.6 consideration.

**Review by jordigilh — COMMENTED** [review]
## Round 2 — Multi-Persona Readiness Re-Audit

**Commits reviewed**: `5c4c8b90b` (batch 1) + `af2560ece` (batch 2)

All 10 findings from round 1 have been verified against the actual code diffs. Results below.

---

### Verification Matrix

| ID | Finding | Resolution | Verified |
|----|---------|------------|----------|
| **SEC-1** | Missing `jwt.WithIssuer()` | Added to `jwt.Parse()` call (`jwt_auth.go:154`) | ✅ |
| **SEC-4** | HTTP JWKS no warning | Startup `WARNING` log with provider/URL fields (`main.go:1142-1146`) | ✅ |
| **SEC-5** | Silent groups fallback | `V(1)` log with username, claim path, error (`jwt_auth.go:173-174`) | ✅ |
| **SEC-6** | DD claims per-provider; code is global | DD pseudocode rewritten to match `NewJWTAuthenticator` actual behavior. Explicitly documents global disable for v1.5, per-provider degradation for v1.6 | ✅ |
| **OPS-4** | Dead `JWTInteractiveGroup` | Removed from `config.go`, Helm template, `values.yaml`, `values.schema.json`, and test files | ✅ |
| **DOC-2** | DD timeout 10s vs code 15s | All three DD references updated (pseudocode, Consequences, Risks table) to 15s. Pseudocode replaced with actual `NewJWTAuthenticator` call pattern | ✅ |
| **DOC-3** | DD per-provider vs global | DD now correctly states: _"If **any** provider's JWKS endpoint is unreachable at startup, `NewJWTAuthenticator` returns an error and the `CompositeAuthenticator` is not created — Pattern B is disabled **globally**."_ | ✅ |
| **QE-3** | TEST_PLAN drift | File paths corrected to `jwt_middleware_test.go`. IT-003/006/008/009 marked `Deferred (v1.6)` with rationale. R4 risk table updated. | ✅ |
| **QE-5/6/7/8** | Missing edge-case tests | 4 new tests added and verified in diff: JWKS cache resilience (QE-5), 20-goroutine concurrency (QE-6), `nbf` enforcement (QE-7), array `aud` (QE-8) | ✅ |
| **QE-9** | PoC vs UT naming | Full `ValidateTokenFull` test added as `UT-KA-1009-004` with `resource_access.kubernaut-agent.roles` double-nested Keycloak claims — verifies username, groups, and ProviderType end-to-end | ✅ |

### Additional observations (non-blocking)

1. **`NewJWTAuthenticator` signature change**: Now accepts `logr.Logger` — all call sites updated (`main.go`, 5 test files). Backward-compatible within repo.
2. **SEC-7** (JWT errors → 500 vs 401): Acknowledged as v1.6 work. Acceptable.
3. **QE-4** (E2E-KA-JWT-005 naming): Acknowledged as cosmetic. Acceptable.
4. **Test count growth**: 22 → 27 JWT-specific unit tests. Good coverage improvement.

---

### Updated Scorecard

| Persona | Round 1 | Round 2 | Notes |
|---------|---------|---------|-------|
| **ProdSec** | PASS | ✅ PASS | SEC-1 hardened, SEC-4 warned, SEC-5 logged |
| **QE** | CONDITIONAL | ✅ PASS | TEST_PLAN aligned, 5 new tests, deferred ITs documented |
| **Product** | PASS | ✅ PASS | No change |
| **Operator/Deployer** | PASS | ✅ PASS | Dead config removed, HTTP warning added |
| **Release Engineering** | PASS | ✅ PASS | 7 commits, clean build |
| **Documentation** | CONDITIONAL | ✅ PASS | DD-AUTH-MCP-001 now matches code |
| **SRE** | PASS | ✅ PASS | No change |

**Verdict: All personas PASS. No remaining blockers. Ready for merge.**
